### DISCUSSION TITLE
Iterate through root site and subsites libraries

### DISCUSSION THREAD

**Discussion by pratheep-r** [pnp/powershell#3186] [Q&A] [answered]
I am trying to loop through libraries in sub sites and main site. The below script only getting the libraries from main site only. 

$SiteURL = "YourSite"

#Connect to PnP Online
$Creds = Get-Credential
Connect-PnPOnline -Url $SiteURL -Credentials $Creds

#Get all subsites in a SharePoint Online site
$Subsites = Get-PnPSubWeb -Recurse -IncludeRootWeb

foreach ($Site in $Subsites)
{
    Write-Host $Site.Url
    Get-PnPList | Where-Object {$_.BaseType -eq "DocumentLibrary" -and $_.Hidden -eq $False}
}

I'd appreciate some help. Seems I am missing something here.

**Comment by AndersRask** [comment]
You should get all subwebs as long as you specify -Recurse

Maybe you dont have access to the subsites?

The code works as intended for me, but i noticed a small typo though in your where filter: it needs to be $_.BaseType and $_.Hidden ($_ is an alias for $PSItem )

**Reply by pratheep-r** [reply]
Thanks for trying. 
I am not sure if I follow you re: typo. Looks like text window removing "_". Please see the attached image
![IterateSubSiteLib-PnPTest ps1 - Visual Studio Code](https://github.com/pnp/powershell/assets/56027088/b82c4926-5e72-44e0-be65-5a0d9e51a593)

If you see the second screen shot, subsite and nested sub sites looped through. However the libraries are only iterated in root site.
![2023-06-13 12_43_36-IterateSubSiteLib-PnPTest ps1 - Output](https://github.com/pnp/powershell/assets/56027088/350a616b-b50a-41cd-9fa2-c6c4e48cae38)

**Reply by pratheep-r** [reply]
Also I have site collection admin access.

**Reply by AndersRask** [reply, marked as answer]
Ah yes good point!

The Get-PnPList in the loop need to know that you are pointing to a new subweb!
You do this by connecting to the subweb inside the loop. Try something like this:
```
 foreach ($Site in $Subsites)
{
    Write-Host $Site.Url
    Connect-PnPOnline -Url $Site.Url
    Get-PnPList
}
```

**Reply by pratheep-r** [reply]
Fab, spot on - it worked. Thanks for the help
